editoast/src/views/timetable/train_schedule.rs

```diff
@@ -590,10 +608,344 @@ async fn occupancy_blocks(
+                PathfindingResult::Failure(_) => {
+                    // The timetable can have a pathfinding failure for a train schedule
+                    // because the train schedule is not valid.
+                    // We skip it because we don't want to compute the track occupancy for an invalid train schedule.
+                    // It's not a problem because we are just looking for the track occupancy of the operational point for the valid trains.
+                    tracing::info!(train_schedule.id, "pathfinding failed");
+                    continue;
+                }
+            };
+
+            let path_projection = PathProjection::new(track_ranges);
+
+            // Get the positions and the times from the simulation
+            let report_train = match simulation {
+                Response::Success { final_output, .. } => final_output.report_train.clone(),
+                _ => {
+                    tracing::info!(train_schedule.id, "simulation failed");
+                    continue;
+                }
+            };
+
+            let path_item_id = match_path_item_id_with_op_id(
+                &path_item_cache,
+                &train_schedule,
+                &operational_point_id,
+            );
+
+            if let Some((track_section, track_occupancy)) = compute_track_occupancy(
+                path_item_id,
+                path_item_positions,
+                &path_projection,
+                &operational_point_track_offsets,
+                &report_train,
+                &train_schedule,
+            ) {
+                track_occupancy_map
+                    .entry(track_section)
+                    .or_default()
+                    .push(track_occupancy)
+            }
+        }
+        track_occupancy_map
+    };
+    Ok(Json(track_occupancy_map))
+}
+
+fn compute_track_occupancy_with_path_item_id(
+    train_schedule: &models::TrainSchedule,
+    path_item_id: &str,
+    path_item_positions: &[u64],
+    path_projection: &PathProjection,
+    operational_point_track_offsets: &[TrackOffset],
+    report_train: &ReportTrain,
+) -> Option<(String, TrackOccupancy)> {
+    let schedule_item = train_schedule
+        .schedule
+        .iter()
+        .find(|schedule| schedule.at.0 == path_item_id);
+    // Stop time at the operational point:
+    // Retrieve the stop time if it exists; otherwise, default to 0
+    let stop_duration = match schedule_item {
+        Some(schedule_item) => schedule_item.stop_for.clone().unwrap_or_default(),
+        // When the path item is the departure position without being a schedule point
+        None => PositiveDuration::default(),
+    };
+
+    // Get the index of the path item in the list of path items that corresponds to the operational point
+    let index = train_schedule
+        .path
+        .iter()
+        .position(|path| path.id.0 == path_item_id)
+        .expect("Unexcpected path item");
+    let position = path_item_positions[index];
```

Comment:
Indexing into an array that comes from the outside of the function (so we can't know the size of it), with an index calculated in the function seems like a recipe for `panic`. But this might due to our design: the pathfinding result is expected to be of the same size with `train_schedule.path` , but its size is non-constrained (by type and compilation I mean).

Reply:
`path_item_positions` is supposed to be the same size with `train_schedule.path`. I understand what do you mean and to be honest I agree. I suggest to add an assert in the pathfinding to compare path_item_positions ond` train_schedule.path `(I don't know if this assert exists in core) ?